src/server/tiering/small_bins.cc

```diff
@@ -109,22 +115,67 @@ std::optional<SmallBins::BinId> SmallBins::Delete(DbIndex dbid, std::string_view
   return std::nullopt;
 }
 
-std::optional<DiskSegment> SmallBins::Delete(DiskSegment segment) {
-  segment = segment.FillPages();
-  if (auto it = stashed_bins_.find(segment.offset); it != stashed_bins_.end()) {
-    stats_.total_stashed_entries--;
-    if (--it->second == 0) {
+SmallBins::BinInfo SmallBins::Delete(DiskSegment segment) {
+  auto full_segment = segment.FillPages();
+  if (auto it = stashed_bins_.find(full_segment.offset); it != stashed_bins_.end()) {
+    stats_.stashed_entries_cnt--;
+    auto& bin = it->second;
+
+    DCHECK_LE(segment.length, bin.bytes);
+    bin.bytes -= segment.length;
+
+    if (--bin.entries == 0) {
+      DCHECK_EQ(bin.bytes, 0u);
       stashed_bins_.erase(it);
-      return segment;
+      return {full_segment, false /* fragmented */, true /* empty */};
+    }
+
+    if (bin.bytes < kPageSize / 2) {
```

Comment:
is it the condition to trigger fragmentation?

Reply:
Yes 🤓 Not very sophisticated